File: NapiJavaWrapper/src/com/nymi/api/wrapper/NymiApi.java
package com.nymi.api.wrapper;

// Singleton manager for NAPI object
import com.nymi.api.LibNapi;

import java.util.AbstractMap;
import java.util.List;


public class NymiApi {

	public enum HapticNotification {
		ERROR(-1),
		NOTIFY_NEGATIVE(0),
		NOTIFY_POSITIVE(1);
		private final int val;
		private HapticNotification(int v) { val = v; }
		public int getVal() { return val; }
	};

	public enum FoundStatus {
		ERROR,
		ANONYMOUS,
		AUTHENTICATED,
		IDENTIFIED,
		PROVISIONING,
		UNCLASPED,
		UNDETECTED,
		UNPROVISIONABLE,
		DISCOVERED
	};
	
	public enum PresenceStatus {
		ERROR,
		DEVICE_PRESENCE_NO,
		DEVICE_PRESENCE_UNLIKELY,
		DEVICE_PRESENCE_LIKELY,
		DEVICE_PRESENCE_YES
	};
	
	public enum ProximityState {
		ERROR,
		PROXIMITY_STATE_NOT_READY,
		PROXIMITY_STATE_UNDETECTABLE, 
		PROXIMITY_STATE_DETECTABLE, 
		PROXIMITY_STATE_SPHERE1,
		PROXIMITY_STATE_SPHERE2, 
		PROXIMITY_STATE_SPHERE3, 
		PROXIMITY_STATE_SPHERE4
	};
	
	public enum KeyType {
		ERROR,
		SYMMETRIC,
		TOTP
	};
	
	
	public enum ProvisionListType {
		ALL,
		PRESENT
	};
	
	
	private static NymiApi instance = null;
	private Listener listener = null;
	
	
	public static NymiApi getInstance() {
		if(instance == null) {
			instance = new NymiApi();
		}
		return instance;
	}
	
    public LibNapi.ConfigOutcome init(NapiReturnObject retobj, String rootDirectory, LibNapi.LogLevel log, int nymulatorPort, String nymulatorHost)
    {
    	LibNapi.ConfigOutcome initResult = LibNapi.INSTANCE.jsonNapiConfigureD(rootDirectory, log, nymulatorPort, nymulatorHost);

    	if (initResult == LibNapi.ConfigOutcome.okay)
    	{
    		listener = new Listener(retobj);
    	}
    	
    	return initResult;
    }
	
	void startProvisioning()
	{
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.start_prov());
	}
	
	void acceptPattern(String pattern)
	{
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.accept_pattern(pattern));		
	}
	
	void stopProvisioning()
	{
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.stop_prov());
	}
	
	void getProvisions(ProvisionListType type)
	{
	    String exchange = (type == ProvisionListType.ALL) ? "provisions" : "provisionsPresent";
	    LibNapi.INSTANCE.jsonNapiPutD(GenJson.get_info(exchange));
	}
	
    void enableOnFoundChange()
    {
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.enable_notification(true,"onFoundChange"));
    }
    
    void enableOnPresenceChange()
    {
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.enable_notification(true,"onPresenceChange"));
    	
    }

    void disableOnFoundChange()
    {
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.enable_notification(false,"onFoundChange"));
    }
    
    void disableOnPresenceChange()
    {
		LibNapi.INSTANCE.jsonNapiPutD(GenJson.enable_notification(false,"onPresenceChange"));
    	
    }
    
    void getApiNotificationState()
    {
    	if (listener != null) {
			LibNapi.INSTANCE.jsonNapiPutD(GenJson.get_state_notifications());
    	}
    }
}

File: Cplusplus/NapiCppWrapper/src/NymiApi.cpp
#include <iostream>
#include "NymiApi.h"
#include "GenJson.h"
#include "Listener.h"

NymiApi *NymiApi::nApi = nullptr;

NymiApi * NymiApi::getNymiApi(nymi::ConfigOutcome &initResult, errorCallback onError, std::string rootDirectory, nymi::LogLevel log, int nymulatorPort, std::string nymulatorHost) {
    
    if (!onError) throw "onError callback is invalid\n";
    
    if (nApi == nullptr) {
        
        nApi = new NymiApi;
        nApi->init(initResult, rootDirectory, log, nymulatorPort, nymulatorHost);

        if (initResult != nymi::ConfigOutcome::okay) {
            delete nApi;
            nApi = nullptr; //caller can call getNymiApi again to reattempt initialization
            return nApi;
        }

        PrivateListener::setOnError(onError);
    }
    
    return nApi;
}

NymiApi::~NymiApi() {

    if (listener.joinable()) {
        PrivateListener::setQuit(true);
        listener.join();                    //must be joined before calling napiTerminate
        nymi::jsonNapiTerminate();

        std::cout << "NymiApi terminated\n";
    }

    nApi = nullptr; //in case we call NymiApi::getNymiApi and need to re-initialize Napi again.
};

//public functions
//----------------
void NymiApi::init(nymi::ConfigOutcome &initResult, std::string rootDirectory, nymi::LogLevel log, int nymulatorPort, std::string nymulatorHost) {
	
	initResult = nymi::jsonNapiConfigure(rootDirectory, log, nymulatorPort, nymulatorHost);

    if (initResult == nymi::ConfigOutcome::okay) {
        listener = std::thread(PrivateListener::waitForMessage);
    }
}

bool NymiApi::startProvisioning(agreementCallback onAgree, newProvisionCallback onProvision) {

    if (!onAgree || !onProvision) return false;
    
    PrivateListener::setOnAgreement(onAgree);
    PrivateListener::setOnProvision(onProvision);
    nymi::jsonNapiPut(start_prov());
    return true;
}

void NymiApi::acceptPattern(std::string pattern) {

	nymi::jsonNapiPut(accept_pattern(pattern));
}

void NymiApi::stopProvisioning() {

	nymi::jsonNapiPut(stop_prov());
}

bool NymiApi::getProvisions(getProvisionsCallback getProvList, ProvisionListType type) {

    if (!getProvList) return false;

    PrivateListener::setProvisionList(getProvList);

    std::string exchange = type == ProvisionListType::ALL ? "provisions" : "provisionsPresent";
    nymi::jsonNapiPut(get_info(exchange));
    return true;
}

bool NymiApi::setOnProvisionModeChange(onStartStopProvisioning onProvisionModeChange){
    
    if (!onProvisionModeChange) return false;
    PrivateListener::setOnProvisionModeChange(onProvisionModeChange);
    return true;
}

bool NymiApi::setOnFoundChange(onNymiBandFoundStatusChange onFoundChange) {

    if (!onFoundChange) return false;
    
    nymi::jsonNapiPut(enable_notification(true, "onFoundChange"));
    PrivateListener::setOnFoundChange(onFoundChange);
    return true;
}

bool NymiApi::setOnPresenceChange(onNymiBandPresenceChange onPresenceChange){
    
   	if (!onPresenceChange) return false;
    
    nymi::jsonNapiPut(enable_notification(true, "onPresenceChange"));
    PrivateListener::setOnPresenceChange(onPresenceChange);
    return true;
}

void NymiApi::disableOnFoundChange(){
    
    nymi::jsonNapiPut(enable_notification(false,"onFoundChange"));
}

void NymiApi::disableOnPresenceChange(){
    
    nymi::jsonNapiPut(enable_notification(false,"onPresenceChange"));
}

bool NymiApi::getApiNotificationState(onNotificationsGetState onNotificationsGet){
    
    if (!onNotificationsGet) return false;
    
    PrivateListener::setOnNotificationsGet(onNotificationsGet);
    nymi::jsonNapiPut(get_state_notifications());
    return true;
}